silkworm/core/execution/processor.cpp

```diff
@@ -70,16 +70,40 @@ void ExecutionProcessor::execute_transaction(const Transaction& txn, Receipt& re
     const intx::uint256 data_gas_price{evm_.block().header.data_gas_price().value_or(0)};
     state_.subtract_from_balance(*txn.from, txn.total_data_gas() * data_gas_price);
 
-    const intx::uint128 g0{protocol::intrinsic_gas(txn, rev, evm_.get_eos_evm_version(), evm_.get_gas_params())};
+    const auto eos_evm_version = evm_.get_eos_evm_version();
+    const auto& gas_params = evm_.get_gas_params();
+
+    const intx::uint128 g0{protocol::intrinsic_gas(txn, rev, eos_evm_version, gas_params)};
     assert(g0 <= UINT64_MAX);  // true due to the precondition (transaction must be valid)
 
     const CallResult vm_res{evm_.execute(txn, txn.gas_limit - static_cast<uint64_t>(g0))};
 
-    const uint64_t gas_used{txn.gas_limit - refund_gas(txn, vm_res.gas_left, vm_res.gas_refund)};
+    uint64_t gas_used{0};
+    if(eos_evm_version < 3) {
+        gas_used = txn.gas_limit - refund_gas(txn, vm_res.gas_left, vm_res.gas_refund);
 
-    // award the fee recipient
-    const intx::uint256 price{evm_.config().protocol_rule_set == protocol::RuleSetType::kTrust ? effective_gas_price : txn.priority_fee_per_gas(base_fee_per_gas)};
-    state_.add_to_balance(evm_.beneficiary, price * gas_used);
+        // award the fee recipient
+        const intx::uint256 price{evm_.config().protocol_rule_set == protocol::RuleSetType::kTrust ? effective_gas_price : txn.priority_fee_per_gas(base_fee_per_gas)};
+        state_.add_to_balance(evm_.beneficiary, price * gas_used);
+    } else {
+        uint64_t storage_gas_consumed{vm_res.storage_gas_consumed};
+        const bool contract_creation{!txn.to};
+        if(contract_creation) {
+            storage_gas_consumed += gas_params.G_txcreate; //correct storage gas consumed to account for initial G_txcreate storage gas
```

Comment:
why it is added here?

Reply:
At the time we deduct the intrinsic gas from the `gas left`, there is no provision to specify that part of the consumed gas (in the case of contract creation) was for storage. Therefore, we handle this after processing the transaction, where we receive an evmc_result.